Docs/types.md

```diff
@@ -0,0 +1,645 @@
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// SessionSecrets list the secrets to use to sign/encrypt and authenticate/decrypt created sessions.
+type SessionSecrets struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Secrets is a list of secrets
+	// New sessions are signed and encrypted using the first secret.
+	// Existing sessions are decrypted/authenticated by each secret until one succeeds. This allows rotating secrets.
+	Secrets []SessionSecret `json:"secrets"`
+}
+
+// SessionSecret is a secret used to authenticate/decrypt cookie-based sessions
+type SessionSecret struct {
+	// Authentication is used to authenticate sessions using HMAC. Recommended to use a secret with 32 or 64 bytes.
+	Authentication string `json:"authentication"`
+	// Encryption is used to encrypt sessions. Must be 16, 24, or 32 characters long, to select AES-128, AES-
+	Encryption string `json:"encryption"`
+}
+
+// IdentityProvider provides identities for users authenticating using credentials
+type IdentityProvider struct {
+	// Name is used to qualify the identities returned by this provider
+	Name string `json:"name"`
+	// UseAsChallenger indicates whether to issue WWW-Authenticate challenges for this provider
+	UseAsChallenger bool `json:"challenge"`
+	// UseAsLogin indicates whether to use this identity provider for unauthenticated browsers to login against
+	UseAsLogin bool `json:"login"`
+	// MappingMethod determines how identities from this provider are mapped to users
+	MappingMethod string `json:"mappingMethod"`
+	// Provider contains the information about how to set up a specific identity provider
+	Provider runtime.RawExtension `json:"provider"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// BasicAuthPasswordIdentityProvider provides identities for users authenticating using HTTP basic auth credentials
+type BasicAuthPasswordIdentityProvider struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// RemoteConnectionInfo contains information about how to connect to the external basic auth server
+	RemoteConnectionInfo `json:",inline"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AllowAllPasswordIdentityProvider provides identities for users authenticating using non-empty passwords
+type AllowAllPasswordIdentityProvider struct {
+	metav1.TypeMeta `json:",inline"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DenyAllPasswordIdentityProvider provides no identities for users
+type DenyAllPasswordIdentityProvider struct {
+	metav1.TypeMeta `json:",inline"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// HTPasswdPasswordIdentityProvider provides identities for users authenticating using htpasswd credentials
+type HTPasswdPasswordIdentityProvider struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// File is a reference to your htpasswd file
+	File string `json:"file"`
```

Comment:
Need to allow user to provide a file that will be accessible by the api.
Uses:
1.
2. As a backup
File
+ likely fits with how users are using this (assuming they have some process that sync a file from idp in some way)
+ easier upgrade
- upgrades might break it
- more than one place to configure

Reference a secret?
+ in line with move away from files of disk
? can we checkpoint this secret for upgrades?

Have a url that specifies whether a file or a secret?

CSI?